scripts/tiling/stitching.py

```diff
@@ -0,0 +1,256 @@
+""" Script for creating a stitched ROI MIBItiff from a folder of FOVs
+"""
+
+import numpy as np
+import os
+from mibidata import mibi_image as mi
+from mibitracker.request_helpers import MibiRequests
+from collections import OrderedDict
+import time
+from mibidata import tiff
+import json
+
+MAX_TRIES = 10
+
+
+def combine_entity_by_name(roi_fov_paths, cols, rows, enforce_square):
+    min_col = np.min(cols)
+    min_row = np.min(rows)
+    cols=[v-min_col+1 for v in cols]
+    rows=[v-min_row+1 for v in rows]
+    w = np.max(cols)
+    h = np.max(rows)
+
+    panel = None
+    out_img = None
+    for fov_i,roi_fov_path in enumerate(roi_fov_paths):
+        print(fov_i+1,roi_fov_path)
+
+        fov_img = tiff.read(roi_fov_path)
+        panel = fov_img.channels
+        fov_img = fov_img.data
+
+        img_shape = list(np.shape(fov_img))
+        ch_count = np.min(img_shape)
+        shape_2d = img_shape[:-1]
+
+        if out_img is None:
+            out_img = np.zeros((h*shape_2d[0], w*shape_2d[1], ch_count), dtype=fov_img.dtype)
+            out_img_shape = list(np.shape(out_img))
+
+        for ch_i in range(ch_count):
+            out_img[((rows[fov_i]-1)*shape_2d[0]):(rows[fov_i]*shape_2d[0]), 
+                    ((cols[fov_i]-1)*shape_2d[1]):(cols[fov_i]*shape_2d[1]), ch_i] = fov_img[:,:,ch_i]
+    
+    if enforce_square:
+        if out_img_shape[0] > out_img_shape[1]:
+            out_img = np.pad(out_img,((0,0),(0,out_img_shape[0]-out_img_shape[1]),(0,0)))
+        elif out_img_shape[0] < out_img_shape[1]:
+            out_img = np.pad(out_img,((0,out_img_shape[1]-out_img_shape[0]),(0,0),(0,0)))
+        
+    return out_img.astype(out_img.dtype, copy=False), panel, int(max(w*shape_2d[1], h*shape_2d[0]))
+
+
+def run_task(fov_paths, out_path, session_dict, mt_upload):
+    unique_rois = np.unique([os.path.basename(p).split("-")[-1].split(".tiff")[0].split("_")[0] for p in fov_paths])
+    roi_path_groups = dict([(u,[p for p in fov_paths if u in p]) for u in unique_rois])
+    mr = None
+    for t in range(MAX_TRIES):
+        try:
+            mr = MibiRequests(**session_dict)
+            break
+        except:
+            if t < MAX_TRIES-1:
+                time.sleep(0.50)
+            else:
+                mr = MibiRequests(**session_dict)
+
+    for roi,roi_fov_paths in roi_path_groups.items():
+        print(roi)
+
+        cols = []
+        rows = []
+        for fov_path in roi_fov_paths:
+            fov_name = os.path.basename(fov_path).split("-")[-1].split(".tiff")[0]
+            c, r = fov_name.split("_")[1:]
+            cols.append(int(c[1:].lstrip("0")))
+            rows.append(int(r[1:].lstrip("0")))
+
+        out_img, panel, max_dim = combine_entity_by_name(roi_fov_paths, cols, rows, enforce_square=True)
```

Comment:
MIBIcontrol allows users to specify a positive or negative spacing between tiles, but this script seems to assume no padding or overlap between adjacent tiles. I think that should be included when reconstructing since it's a parameter that's available to set and the value (can be specified for each the x- and y-dimension) is available in the run JSON file.

Reply:
Adding this in would take me some time, as I have never looked at that parameter (nor was aware of it). The situation where we add spacing is more straightforward, but a negative spacing would need some decisions to be made regarding how it's all handled: Do we blend them all, have the first FOV's value take precedence, or the last?

Reply:
I think we do need to include it. That said, the intent of this script is to be a usable tool for users but more importantly to be a starting point reference for users who want/need to do something more complex. So, we don't necessarily need to solve the problem (specifically for overlapping tiles) in a robust manner. A simple solution may be all that's needed here.

For the negative spacing, it can be as simple as pixels being overwritten as the image is stitched together. If you iterate through the y tiles and then the x tiles from 0 to n, then the pixels will be overwritten in a predictable way. And that should suffice for the purpose of a working example.

For the actual spacing/overlap parameter, you could take one of two approaches.
1. Load the run JSON file that's in the folder with all the TIFFs and use the `xMargin` and `yMargin` values. The downside here is that MIBIcontrol defines them in microns, but you need pixels. So, you'd need to read the metadata of one of the MIBItiffs to get the pixel scaling of the images and convert microns to pixels.
2. Add xMargin and yMargin (in pixel units) as arguments to the script or as variables that can be set. They should default to 0, and then a user can override them.